wordpress-mobile/WordPress-Android
# New per app language preferences implementation

This PR adds [Per-app language preferences](https://developer.android.com/about/versions/13/features#app-languages)  introduced in Android 13 (API 33) level.

- Updates compileSDKVersion to 33
- Updates androidxAppCompatVersion to 1.6.0-rc01
- Updates for API changes after upgrading compileSdkVersion to 33
- Let AndroidX handle the locale storage so that the user's preference persists
- Provides migration for handing over storage to system

> **Note**
> This PR is going to be parked until such time `androidxAppCompatVersion` is upgraded to ~ 1.6.0 and `compileSdkVersion` to 33 in separate PRs

TODO:
- [ ] Set allWarningsAsErrors to true before merge (Set it to false to avoid updating a large number files in this PR)
- [ ] Auto generate locales_config.xml using release-toolkit when we download the list of locales from GlotPress.
- [ ] Probably need to upgrade Android Gradle Plugin from 7.2.1 to 7.3.1 or later in settings.gradle

Settings > System > Languages and Input > App languages > [App] > App language

https://user-images.githubusercontent.com/990349/190540086-ec27306b-7149-4621-92af-9923061c6a39.mp4

Settings > Apps > [App] > Language > App language

https://user-images.githubusercontent.com/990349/190540143-37ae77a9-a96c-4552-944f-2b02b8626981.mp4

To test:

Test 1:
- Build and run
- Launch WordPress or Jetpack app on a device running Android 13
- Now go to device **Settings > System > Languages & Input > App Languages** > Select (WordPress app or Jetpack app)
- Ensure WordPress or Jetpack app shows in App Lanaguages
- Tap on the app.  It should navigate to App Lanaguage selection screen

Test 2:
- Build and run
- Launch WordPress or Jetpack app on a device running Android 13
- **Settings > Apps > Select (WordPress app or Jetpack app) > Language** (scroll if required)
- Tap on the Language option.  It should navigate to App Lanaguage selection screen

Test 3
- WordPress app / Jetpack app already have in app Language picker ( Me > App Setttings > Interface Language)
- Select another language
- Ensure, the language change take effect
- Navigate to **Settings > Apps > Select (WordPress app or Jetpack app) > Language** (scroll if required)
- Tap on the Language option.  It should navigate to App Lanaguage selection screen
- Ensure, the language selected in in-app language picker also takes effect in System settings

Test 4
- Build and run
- Launch WordPress or Jetpack app on a device/emulator running **Android 12**
- Navigate to in app Language picker ( Me > App Setttings > Interface Language)
- Select another language
- Ensure, the language change takes effect

Test 5
- Build and run
- Launch WordPress or Jetpack app on a device/emulator running **Android 12**
- Upgrade to **Android 13**
- Navigate to in app Language picker ( Me > App Setttings > Interface Language), Note language
- Navigate to **Settings > Apps > Select (WordPress app or Jetpack app) > Language** (scroll if required)
- Tap on the Language option.  It should navigate to App Lanaguage selection screen, Note language
- Ensure, the language selected in in-app language picker, and in Systems settings are same

> **Note**
> Need to evaluate when we can potentially remove in-app language picker altogether.  Potentially when **targetSdkVersion** or **minSdkVersion** is upgraded to Android 13 (API 33).



To test:

## Regression Notes
1. Potential unintended areas of impact
Language and Locale in some places

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Tested various features and in both ways from in-app picker to system settings and vice versa

3. What automated tests I added (or what prevented me from doing so)
Existing unit tests

PR submission checklist:
